core/src/main/java/hudson/Util.java

```diff
@@ -179,28 +179,57 @@ public static String replaceMacro(@CheckForNull String s, @Nonnull VariableResol
     }
 
     /**
-     * Loads the contents of a file into a string.
+     * Reads the entire contents of the text file at <code>logfile</code> into a
+     * string using the {@link Charset#defaultCharset() default charset} for
+     * decoding. If no such file exists, an empty string is returned.
+     * @param logfile The text file to read in its entirety.
+     * @return The entire text content of <code>logfile</code>.
+     * @throws IOException If <code>logfile</code> cannot be converted to a
+     * {@link java.nio.file.Path} or if an error occurs while reading the file.
+     * @deprecated call {@link #loadFile(java.io.File, java.nio.charset.Charset)}
+     * instead to specify the charset to use for decoding (preferably
+     * {@link java.nio.charset.StandardCharsets#UTF_8}).
      */
     @Nonnull
+    @Deprecated
     public static String loadFile(@Nonnull File logfile) throws IOException {
         return loadFile(logfile, Charset.defaultCharset());
     }
 
+    /**
+     * Reads the entire contents of the text file at <code>logfile</code> into a
+     * string using <code>charset</code> for decoding. If no such file exists,
+     * an empty string is returned.
+     * @param logfile The text file to read in its entirety.
+     * @param charset The charset to use for decoding the bytes in <code>logfile</code>.
+     * @return The entire text content of <code>logfile</code>.
+     * @throws IOException If <code>logfile</code> cannot be converted to a
+     * {@link java.nio.file.Path} or if an error occurs while reading the file.
+     */
     @Nonnull
     public static String loadFile(@Nonnull File logfile, @Nonnull Charset charset) throws IOException {
-        if(!logfile.exists())
-            return "";
-
-        StringBuilder str = new StringBuilder((int)logfile.length());
+        return loadFile(fileToPath(logfile), charset);
+    }
 
-        try (BufferedReader r = Files.newBufferedReader(fileToPath(logfile), charset)) {
-            char[] buf = new char[1024];
-            int len;
-            while ((len = r.read(buf, 0, buf.length)) > 0)
-                str.append(buf, 0, len);
+    /**
+     * Reads the entire contents of the text file at <code>path</code> into a
+     * string using <code>charset</code> for decoding. If no such file exists,
+     * an empty string is returned.
+     * @param path Path to the text file to read in its entirety.
+     * @param charset The charset to use for decoding the bytes in the file.
+     * @return The entire text content of the file.
+     * @throws IOException If an error occurs while reading the file.
+     * @since TODO
+     */
+    @Nonnull
+    public static String loadFile(@Nonnull Path path, @Nonnull Charset charset) throws IOException {
+        byte[] bytes;
+        try {
+            bytes = Files.readAllBytes(path);
+        } catch (NoSuchFileException ignored) {
+            return "";
```

Comment:
It is unclear to me why this is preferable over just wrapping it in an `IOException` that already gets thrown.

Reply:
The existing code initialized the capacity of the StringBuilder to `(int)logfile.length()` outside the try-with-resources statement that opened the file.

It is possible to move the initialization of the StringBuilder within the try-with-resources statement with something like:
```java
@Nonnull
public static String loadFile(@Nonnull File logfile, @Nonnull Charset charset) throws IOException {
    StringBuilder str;

    try (BufferedReader r = Files.newBufferedReader(fileToPath(logfile), charset)) {
        str = new StringBuilder((int)logfile.length());

        char[] buf = new char[1024];
        int len;
        while ((len = r.read(buf, 0, buf.length)) > 0)
            str.append(buf, 0, len);
    } catch (NoSuchFileException e) {
        return "";
    }

    return str.toString();
}
```

However, I looked at the Oracle JDK implementation of readAllBytes() and I thought that it was better at handling special cases such as avoiding casting a too-large `long` value to `int`, the file being truncated while it is read, or the file being written to as it is being read:

```java
/**
 * The maximum size of array to allocate.
 * Some VMs reserve some header words in an array.
 * Attempts to allocate larger arrays may result in
 * OutOfMemoryError: Requested array size exceeds VM limit
 */
private static final int MAX_BUFFER_SIZE = Integer.MAX_VALUE - 8;

private static byte[] read(InputStream source, int initialSize) throws IOException {
    int capacity = initialSize;
    byte[] buf = new byte[capacity];
    int nread = 0;
    int n;
    for (;;) {
        // read to EOF which may read more or less than initialSize (eg: file
        // is truncated while we are reading)
        while ((n = source.read(buf, nread, capacity - nread)) > 0)
            nread += n;

        // if last call to source.read() returned -1, we are done
        // otherwise, try to read one more byte; if that failed we're done too
        if (n < 0 || (n = source.read()) < 0)
            break;

        // one more byte was read; need to allocate a larger buffer
        if (capacity <= MAX_BUFFER_SIZE - capacity) {
            capacity = Math.max(capacity << 1, BUFFER_SIZE);
        } else {
            if (capacity == MAX_BUFFER_SIZE)
                throw new OutOfMemoryError("Required array size too large");
            capacity = MAX_BUFFER_SIZE;
        }
        buf = Arrays.copyOf(buf, capacity);
        buf[nread++] = (byte)n;
    }
    return (capacity == nread) ? buf : Arrays.copyOf(buf, nread);
}

public static byte[] readAllBytes(Path path) throws IOException {
    try (SeekableByteChannel sbc = Files.newByteChannel(path);
         InputStream in = Channels.newInputStream(sbc)) {
        long size = sbc.size();
        if (size > (long)MAX_BUFFER_SIZE)
            throw new OutOfMemoryError("Required array size too large");

        return read(in, (int)size);
    }
}
```

I also liked the use of SeekableByteChannel.size() instead of File.length().

Unfortunately, there isn't a more specialized standard API for reading all characters in a file, say Files.readAllChars(Path, Charset), so I have resorted to calling Files.readAllBytes() to read all bytes, and then constructing the String from the bytes and specified charset.

I don't know this project's philosophy on whether it is better to reuse existing code or replicate and specialize code for each particular case. If the latter, I could add another commit that would do what a hypothetical Files.readAllChars(Path, Charset) API would do.